sdk/python/feast/pyspark/historical_feature_retrieval_job.py

```diff
@@ -270,14 +270,19 @@ def join_entity_to_feature_tables(
     return joined
 
 
-def retrieve_historical_features(spark: SparkSession, conf: Dict) -> DataFrame:
+def retrieve_historical_features(
+    spark: SparkSession, conf: Dict, entity_df: DataFrame = None
+) -> DataFrame:
     """Retrieve batch features based on given configuration.
 
     Args:
         spark (SparkSession):
             Spark session.
         conf (Dict):
-            Configuration for the retrieval job, in json format. Sample configuration as follows:
+            Configuration for the retrieval job, in json format.
+        entity_df (DataFrame):
+            Optional. If provided, the entity will be used directly and conf["entity"] will be ignored.
```

Comment:
What does `entity will be used directly` mean. Not clear why `entity_df` can be optional.

Reply:
It's mainly to support the scenario below:
`client.get_historical_features_df(["rating"], entity_df)`

Where a user passed in a Spark dataframe / Pandas dataframe directly as entity, instead of using `FileSource` , `BigQuerySource`.

If we mandate user to always use `FileSource` or `BigQuerySource`, then we don't need this additional argument.

Reply:
I dont understand why its part of this method. Doesnt the entity_df get uploaded prior to calling this method, meaning it should be configured through the conf?